### ISSUE TITLE
Regarding the Infinite Water source block

### ISSUE THREAD

**Issue by HipHopHuman** [TeamPneumatic/pnc-repressurized#670]
I was having a read of the roadmap and I read the part about the infinite water source block. I like the idea, but from my experience playing other mods that add their own version of a water source block (Nuclearcraft's Infinite Water Source, Thermal's Acqeous Accumulator, Cooking For Blockheads' Kitchen Sink...) I've found that these blocks tend to feel very plain, and offer far too much utility for the investment they require. Using vanilla to bucket in water via dispensers or a liquid hopper restricts the water generation to a specific rate per tick, so you have to build infrastructure around it, which is fun. The other mods don't have these restrictions, so it's a rather boringly straightforward task.

The way the water source block is described makes it sound like it behaves similarly to the Aqceuous Accumulator, which seems out of place in PneumatiCraft.

I was thinking about this for a while, and this seems like a job better suited for the Gas Lift. It's already a mechanic in the mod used for pumping liquids, which is essentially what this Infinite Water source would be doing.

What do you think about applying positive pressure to a gas lift so that it drills until it hits bedrock, then reversing the pressure into the negative to obtain infinite water? This makes more sense to me, as a PNC mechanic, and makes infinite water generation feel a little more balanced.

**Comment by HipHopHuman** [comment]
I was doing a bit of research into uses for compressed air in the real world and came across a machine called the [Trompe](https://en.wikipedia.org/wiki/Trompe), a giant machine that generates compressed air from falling water. Typically these machines were used to power Catalan Forges, a type of Bloomery Forge. This feels a bit coincedantal, considering you're looking for a collective name for your water-powered ore processing system :) Bloomeries are a bit of a primitive way of processing ore, so I'm not sure if the name fits, but worth a look.

**Comment by desht** [comment]
> The way the water source block is described makes it sound like it behaves similarly to the Aqceuous Accumulator, which seems out of place in PneumatiCraft.

Yeah, that's part of the reason why it's not in yet (that, and I haven't really come up with a truly satisfying way of implementing a smelting system yet...).

> What do you think about applying positive pressure to a gas lift so that it drills until it hits bedrock, then reversing the pressure into the negative to obtain infinite water? This makes more sense to me, as a PNC mechanic, and makes infinite water generation feel a little more balanced.

Thing about this is, it's a lot simpler (and works today) to just put a gas lift over a 2x2 pool of water and give it a very small amount of pressure (0.5 bar is fine if the water is directly underneath), and that's it.  The main reason I wanted a single-block infinite water source is to minimise lag from frequent water block updates, since my idea for the smeltery would require a large amount of water in the ore-washing stage.

**Comment by HipHopHuman** [comment]
I'll reply to this backwards...

> Thing about this is, it's a lot simpler (and works today) to just put a gas lift over a 2x2 pool of water and give it a very small amount of pressure (0.5 bar is fine if the water is directly underneath), and that's it. The main reason I wanted a single-block infinite water source is to minimise lag from frequent water block updates, since my idea for the smeltery would require a large amount of water in the ore-washing stage.

Yes, it is a lot simpler to do that, but you're still somewhat restricted by the speed at which vanilla refreshes the water source block that the Gas Lift is pumping from. What I'm suggesting here is to simply add a special case for the Drill Pipe encountering bedrock, where it will just generate water from thin air:tm: and pretend that it's coming "from beneath the bedrock" - basically the same way your idea works, but with a few more steps, so the player has the option of infinite water generation, but it isn't as easy as "place block between two lumps of water to win"

> Yeah, that's part of the reason why it's not in yet (that, and I haven't really come up with a truly satisfying way of implementing a smelting system yet...).

Well, looking at your notes it seems you have a lot of familiar mechanics described. It's one thing to have it be thematic with IRL pneumatic ore processing, but I will say I like that it introduces concepts the player is familiar with. Crushers being one, and the 3-tall multiblock being similar in construct to the refinery. Makes it easier to quantify and think about, IMO. That being said, I actually did a little research into pneumatic ore processing systems. Perhaps my findings would interest or inspire you? It's worth noting that while compressed air is used heavily in the mining operation these days, that isn't really true for the ore processing side of things. The most notable use of compressed air in ore processing is in that of a bloomery, which is a rather primitive method of processing ore. It's use is dated back to the late 1800s and early 1900s. I previously mentioned the Trompe, a compressed air generating machine that used the impact of falling water to stoke a bloomery forge. That later got upgraded to a system that used a compressed air powered hammer for shaping ingots, along with a device known as the Catalan Forge. I feel like the design of these devices is a bit too primitive (being mostly made from clay, brick and other ceramics) to fit thematically in PneumaticCraft, which gives off an "It's Iron Man and you're Tony Stark, but with Lego:tm:" vibe.

I did some further research and found a few more modern processes that use pneumatic concepts for ore processing.

The first step is the mining and crushing of ores. Unfortunately there's not much info on how this is done in conjunction with compressed air, so I assume just a generic crushing mechanic is used that uses compressed air as a power source. So a hammer, roller or press. This process typically happens directly in the mine, so the mine itself produces a crushed ore product. This product is then fed into a pneumatic sorting system which is described by this diagram:

![image](https://user-images.githubusercontent.com/70091260/101245212-c7beab00-3713-11eb-89f7-6811045ebdac.png)

I am not exactly sure what this process is called, but essentially, an X-Ray or Laser powered detection system feeds data input into a microchip fitted with a sorting algorithm that adjusts a pneumatic actuator, which changes it's trajectory to sort the waste product (rock) from the mineral (ore) product. This could be facilitated in PneumaticCraft with the use of PCBs and the Air Cannon. 

This mineral product still has some waste byproduct attached, so it is then fed into a Froth Flotation machine, which is essentially a big tank that is injected with pulp and a reagent (like oil or sulphur) that are both respectably hydrophobic/hydrophilic. This water, pulp, ore and reagent mixture is agitated by using compressed air to inject bubbles into the water, with a rotating *impeller* turbine blade at the bottom that creates a vortex that forces the bubbles upwards to separate the metal material product (via the hydrophilic reagent) to the top of the water in the form of froth, or slurry. 

![image](https://user-images.githubusercontent.com/70091260/101245354-fbe69b80-3714-11eb-959e-82de48e8fed6.png)

![image](https://user-images.githubusercontent.com/70091260/101245370-1751a680-3715-11eb-8cb6-7936d7ea9560.png)

![image](https://user-images.githubusercontent.com/70091260/101245384-2afd0d00-3715-11eb-8b32-f84bff1e4410.png)

![image](https://user-images.githubusercontent.com/70091260/101245432-934bee80-3715-11eb-9063-9886ed231a14.png)

So, this machine would require an input of oil, water and some kind of pulp, and would export froth, or Slurry (the stuff you turn into ingots) and Tailing (the waste material, pumped from the bottom of the Flotation machine). I'm not sure if that's a mechanic you're interested in having in PneumaticCraft, but it seems cool to me. Perhaps that tailing could be repurposed in some way, like diverting it to a water treatment facility so the water can be re-used, or somehow turning it into a trace amount of biofuel or something, or just shaping it into andesite, cobble, gravel, granite, diorite or whatever. In the real world, this Tailing contains significant amounts of Limestone. Maybe you could add Limestone as a decorative block, but I would advise against that as many mods do this already (Quark, Strata, Create, just to name a few).

Once the slurry is acquired, it is then transported into an assembly where it is turned into pellets (the process is called Sintering). The pellets are then routed into a blast furnace:

![image](https://user-images.githubusercontent.com/70091260/101245548-58968600-3716-11eb-802c-7b7877e50cd3.png)

Wikipedia describes Sintering as follows:

> Sintering or frittage is the process of compacting and forming a solid mass of material by heat or pressure without melting it to the point of liquefaction. Sintering happens naturally in mineral deposits or as a manufacturing process used with metals, ceramics, plastics, and other materials.

So, what this sounds like in practice to me is the process of drying. I imagine that the slurry output from the Flotation machine is pretty wet, so it goes through a process that uses heat to dry it off, forming a coarse dust. This dust then goes through a mesh filtration system, which basically cleans impurities off of the metal. 

This is somewhat confirmed by this diagram, but in reverse (it filters before the drying process): 

![image](https://user-images.githubusercontent.com/70091260/101245828-46b5e280-3718-11eb-9e61-f9e37c0625e3.png)

Basically, the end result is compressed pellets of highly concentrated metal, which make their way into a blast furnace to be melted down into a liquid where they are then formed into ingots.

This obviously seems like a very involved process, and it would probably be a monumental task to code, with risk of it confusing the player, so I think that if any of this inspires you, it may be worth utilizing your creative discretion to deviate from the way it works in the real world and incorporate some simplifications here and there for gameplay purposes.

What I like about this is that a lot of the fundamental concepts described here already exist in PneumaticCraft in some way, like the Air Cannon, Vortex Tube, Heat system, Gas Lift, Oil, etc, and it would be interesting to combine a lot of these different elements into "the ultimate ore processing system" that is a little more unique in it's realism than the sci-fi 1-block-that-does-some-stuff systems that other mods use. I'm pretty sure Gregtech players would go nuts for the realism, lol.

**Comment by desht** [comment]
Yeah, this sort of complexity pretty much requires its own mod.  Yes, it would be very cool, but also (as you say) a huge coding task, and likely to make the vast majority of players take one look and go running for Thermal or Mek ore processing.

I do want to make something a little different, and I still want to do something with the idea of balancing pressure & power to optimum points (which vary per ore) to get the best output.  Mess up the balancing and you still get a bit of boost (starting at around 1.5x), but get it right, and you get a potentially huge boost (up to 5x, but that should be challenging to maintain).

**Comment by HipHopHuman** [comment]
I mean, if you take these 3 steps and simplify them as much as possible you pretty much end up with the mechanic described in the roadmap anyway xD. First process is crushing, which sends output to big machine that spins water (i.e. ore washing machine) and then off to the blast furnace it goes. I just read about the froth flotation machine and how waste material has to be extracted from it and how reagents had to be added to it to keep it functioning and thought that was cool, and how the pelletization process required heat, but not *too* much heat that the metal is melted. I thought it fit with your idea of maintaining optimum points of temperature to get better output.  

I was thinking of having it be simple as well, but just thematically naming the machines and their processes and maybe basing the models of the machines off of their real-world counterparts so that it fits better with PneumaticCraft's theme, hence:

> it may be worth utilizing your creative discretion to deviate from the way it works in the real world and incorporate some simplifications here and there for gameplay purposes.

**Comment by MuteTiefling** [comment]
Just to weigh in a little, one of the optimizations Mekanism made going to 1.16 was that their water pump no longer removes the source block in world. If the blocks beneath it are a valid "infinite source" formation, they simply generate water internally at an appropriate rate, thereby eliminating the constant updates of making a new water source. 

Seems like that would be a suitable compromise for performance while maintaining the thematic gas lift.

**Comment by desht** [comment]
Yeah, only possible downside is if the world is in some kind of no-infinite-water mode. But I guess it could be controlled with a config setting.

**Comment by HipHopHuman** [comment]
> Yeah, only possible downside is if the world is in some kind of no-infinite-water mode. But I guess it could be controlled with a config setting.

So I actually play with "no infinite water mode" quite often (I like building infrastructure around limited water supplies, it's a fun puzzle - and if you're like me I recommend "fluid-physics"). I think you need not worry, as most finite water mods are included mostly in packs that balance around the mechanic, which means PNC's pumps would probably just get nerfed in favor of some other more end-gamey way of getting large water throughputs.
